lite2/client.go

```diff
@@ -0,0 +1,374 @@
+package lite
+
+import (
+	"bytes"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/lite2/provider"
+	"github.com/tendermint/tendermint/lite2/store"
+	"github.com/tendermint/tendermint/types"
+)
+
+// TrustOptions are the trust parameters needed for when a new light client
+// connects to the network or when a light client that has been offline for
+// longer than the unbonding period connects to the network.
+//
+// The expectation is the user will get this information from a trusted source
+// like a validator, a friend, or a secure website. A more user friendly
+// solution with trust tradeoffs is that we establish an https based protocol
+// with a default end point that populates this information. Also an on-chain
+// registry of roots-of-trust (e.g. on the Cosmos Hub) seems likely in the
+// future.
+type TrustOptions struct {
+	// Only trust commits up to this old.
+	// Should be equal to the unbonding period minus a configurable evidence
+	// submission synchrony bound.
+	Period time.Duration
+
+	// Header's Height and Hash must both be provided to force the trusting of a
+	// particular header.
+	Height int64
+	Hash   []byte
+}
+
+type mode int
+
+const (
+	sequential mode = iota
+	skipping
+)
+
+// Option sets a parameter for the light client.
+type Option func(*Client)
+
+// SequentialVerification option configures the light client to sequentially
+// check the headers. Note this is much slower than SkippingVerification,
+// albeit more secure.
+func SequentialVerification() Option {
+	return func(c *Client) {
+		c.mode = sequential
+	}
+}
+
+// SkippingVerification option configures the light client to skip headers as
+// long as {trustLevel} of the old validator set signed the new header. The
+// bisection algorithm from the specification is used for finding the minimal
+// "trust path".
+//
+// trustLevel - maximum change between two not consecutive headers in terms of
```

Comment:
mmm... faults have nothing to do with trust assumptions (I mean validator set can change because of faults, but not always).

> Given that this is the fraction of faults we will tolerate, not the fraction of the validator set we "trust" (assume is correct)

I think it's neither of those. It's the fraction of the trusted old validator set, which has signed the new commit, required to trust it. I think the existing name captures this pretty well. Although maybe we should rename to `trustThreshold`.

Reply:
It's not 'the fraction of the trusted old validator set', since it's the maximum change (so it's `1 - the fraction of the trusted old validator set`, as it were), right?

Reply:
updated comment to be `fraction of the old validator set (in terms of voting power),
// which must sign the new header in order for us to trust it.`

Reply:
Great - thanks.

Reply:
But I don't think that's what this is - it's `1/3`, not `2/3`, concretely.

See https://github.com/tendermint/tendermint/pull/3989/files#diff-f6c374f81d77f2ae4041a970bc378b89R16

Whereas the "fraction of the old validator set (in terms of voting power) which must sign the new header in order for us to trust it" is `2/3`.

Reply:
```
      // get validators in h1 that signed h2
      // sum of voting powers in h1 of
      // validators that signed h2
      // is more than a third in h1
```

from the spec ^^

Are you trying to say we should stress out that `trustLevel` is only used for non-consecutive headers?

Reply:
Right, you do say `non-consecutive`, I just think clearer wording would help - the idea (as I understand it) is that we assume that `> 2/3` are correct, so that if `> 1/3` sign a new header at least one of the signers will have been in the correct set.